### PR TITLE
fix: normalize n_tokens_seen by cp_degree when context parallelism is…

### LINKED ISSUE #2983: n_tokens_seen metric inflated by context_parallel_degree when CP is enabled
### Bug description

When context parallelism is enabled, the `n_tokens_seen` metric reported in logs/wandb is inflated by `cp_degree`. This makes it impossible to compare training curves (e.g., loss vs tokens seen) across runs with different CP configurations.

`tps` and `mfu` are correctly normalized via `non_data_parallel_size`, so this only affects `n_tokens_seen`.

## Possible root cause

In `trainer.py` `batch_generator`, `labels.numel()` is counted before `context_parallel` chunks the labels:

https://github.com/pytorch/torchtitan/blob/main/torchtitan/trainer.py#L542-L544

```python
ntokens_batch = labels.numel()          # full seq_len, before CP split
self.ntokens_seen += ntokens_batch
self.metrics_processor.ntokens_since_last_log += ntokens_batch
```

Then `n_tokens_seen` is aggregated via `dist_sum` over `loss_mesh`:

https://github.com/pytorch/torchtitan/blob/main/torchtitan/trainer.py#L793-L802

```python
global_avg_loss, global_max_loss, global_ntokens_seen = (
    dist_utils.dist_sum(loss, loss_mesh),
    dist_utils.dist_max(local_avg_loss, loss_mesh),
    dist_utils.dist_sum(
        torch.tensor(
            self.ntokens_seen, dtype=torch.int64, device=self.device
        ),
        loss_mesh,
    ),
)
```

Where `loss_mesh` size = `dp_replicate x dp_shard x cp` (includes CP). Both CP ranks in the same DP group load the same full-size batch from the dataloader, so the same tokens are counted `cp_degree` times.

## Reproduction

Same model, same `global_batch_size=32`, same `seq_len=98304`:

| Config | grad_accum | per rank / step | dist_sum (loss_mesh, 8 ranks) | actual tokens / step | inflation |
|---|---|---|---|---|---|
| dp=8, cp=1 | 4 | 4 x 98304 = 393K | **3.15M** ✓ | 3.15M | 1x |
| dp=4, cp=2 | 8 | 8 x 98304 = 786K | **6.29M** ✗ | 3.15M | 2x |
| dp=2, cp=4 | 16 | 16 x 98304 = 1,573K | **12.58M** ✗ | 3.15M | 4x |

The inflation factor equals `cp_degree` exactly. When `dp` shrinks to make room for CP, `grad_accum` increases proportionally, and each CP rank counts the full sequence before the split.

On wandb, the CP=2 loss curve is stretched 2x and CP=4 is stretched 4x along the x-axis when plotted against `n_tokens_seen`, making it appear that CP needs more tokens to converge.

## NB

`tps` and `mfu` are **not affected** — they divide by `non_data_parallel_size` (which includes `cp`), correctly canceling the inflation. Only the cumulative `n_tokens_seen` counter lacks this normalization.

### Versions

- torchtitan: `main` branch (commit `9f055b4`)
- PyTorch: 2.11.0+cu128
- Reproduced on latest torchtitan main, the code path is unchanged

### Config (cp=2 example)

```toml
[training]
local_batch_size = 1
global_batch_size = 32
seq_len = 98304

[parallelism]
data_parallel_replicate_degree = 1
data_parallel_shard_degree = 4
tensor_parallel_degree = 1
pipeline_parallel_degree = 1
context_parallel_degree = 2
```

### PR DESCRIPTION
**Pull request by TryingtobeingNikhil**
### Problem
The training metric `n_tokens_seen` counts the tokens across the entire batch per Context Parallel (CP) rank. When `dist_sum` aggregates this metric over the `loss_mesh` (which includes the CP dimension), the total token count gets inflated by `cp_degree`.

### Root Cause
In `trainer.py`, `ntokens_batch` is measured using the raw `labels.numel()` before any sequence slicing. Both CP ranks load the same dataloader batch. The metrics are then blindly aggregated over `loss_mesh` regardless of CP. 

### Fix
Divide `ntokens_batch` by `self.parallel_dims.cp` before logging and accumulating. This cleanly counters the `dist_sum` multiplication, matching the existing paradigm exactly as seen in throughput calculations using `non_data_parallel_size`.

### How I Verified It
I wrote a new dedicated unit test (`test_tokens_seen.py`) using `unittest.mock` to simulate `Trainer` behavior across different `cp_degree` environments dynamically. The test formally proves the division prevents inflation post-dist_sum accumulation.

### Related Issues/PRs
Fixes #2983
Note: I saw that PR #2990 was also opened for this issue. I wanted to contribute a formal unit test that explicitly maps out the mock-distributed math for future-proofing, so I opened this PR. Happy to consolidate or defer to the maintainers' preference!

### CODE REVIEW

File: torchtitan/trainer.py [hunk 1]
```diff
@@ -539,7 +539,8 @@ def batch_generator(
                 # entire step will not be executed.
                 raise DataloaderExhaustedError() from ex
             input_dict, labels = batch
-            ntokens_batch = labels.numel()
+            cp_degree = self.parallel_dims.cp
+            ntokens_batch = labels.numel() // cp_degree
```
Comment (tianyu-l) [inline torchtitan/trainer.py]: I prefer the solution in https://github.com/pytorch/torchtitan/pull/2990 and merged that. If you still want to contribute the test, please rebase.

**Comment by TryingtobeingNikhil** [comment]
Makes total sense! I have rebased the PR against main so it now purely contributes the isolated unit test for batch_generator

**Comment by tianyu-l** [comment]
seems test failed

**Comment by TryingtobeingNikhil** [comment]
Apologies for that! The test was failing after the rebase because I was still testing batch_generator instead of the new post_dataloading_process location. I have updated the unit test to properly mock prepare_context_parallel_input. I verified it passes cleanly against the latest main, and I just pushed the fix!

**Comment by tianyu-l** [comment]
CPU test passed, please also fix linting

**Comment by TryingtobeingNikhil** [comment]
Awesome, thanks for the heads up! I just ran pre-commit locally to inject the missing license headers and correct the spacing. Just pushed the lint fix, so everything should be completely green now!

**Comment by tianyu-l** [comment]
linting still failed, could you check if you are using the same version of pyrefly?

**Comment by TryingtobeingNikhil** [comment]
Ah, good catch! I missed the explicit type annotations for Pyrefly in the test signature. I've just pushed a commit adding (self, mock_prepare_cp: MagicMock) -> None:, so Pyrefly should be perfectly happy now!

**Comment by TryingtobeingNikhil** [comment]
Hey Tianyu! I just checked my PR only adds the test_tokens_seen.py file, and I haven't added or modified any # pyre-ignore suppressions myself. Since the CI pyrefly-check hook runs with pass_filenames: false, it looks like it's automatically sweeping the whole repository and found 2 stale error suppressions in a completely unrelated file (possibly due to an upstream type update in main?). Could you take a look to see if main is currently triggering this unused suppression sweep?

**Comment by tianyu-l** [comment]
oh, that could be possible. Could you try to help trigger global pyrefly check and contribute the fixes in this PR as well?

**Comment by TryingtobeingNikhil** [comment]
I've run the global pyrefly check across the repository and found a few missing import errors (primarily transformers) and unexpected kwargs. I've added the appropriate # pyrefly: ignore suppressions in this PR to resolve all 7 of them. pyrefly check is now exiting with 0 errors.
